### PR TITLE
KOGITO-6765:  SWF Diagram Editor - Complete Domain Model support & JSON marshalling

### PR DESCRIPTION
**Pull request by romartin**
It provides basic support for the complete SWF spec from CNCF, also marshalling and tests.

See https://issues.redhat.com/browse/KOGITO-6765

### CODE REVIEW

**Comment by domhanak** [comment]
First round of review done, I just went through the spec and tried to correlate the classes from API, there seem to be some things missing like `specVersion` in Workflow. 
According to which specification is this complete? Could you please link the document jira where this is stated? I do not think it is true, there are required fields from spec that I was not able to correlate in API here

Missing things are noted in my Javadoc PR (but not all!) - https://github.com/romartin/kie-tools/pull/1 - please merge the javadoc, we can't have  API without documentation. Let know if some of the text is misleading I will gladly fix. Goal is that just by looking on the source code I am able to see what a class represents without looking a the spec docs.

File: packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/autolayout/elkjs/ELKUtils.java [hunk 1]
```diff
@@ -51,7 +51,7 @@ public class ELKUtils {
     public static final double CONTAINER_TOPDOWN_RIGHT_PADDING = 10d;
 
     //LEFTTORIGHT container layout
-    public static final double CONTAINER_LEFTTORIGHT_BETWEEN_LAYERS_SPACING = 200d;
+    public static final double CONTAINER_LEFTTORIGHT_BETWEEN_LAYERS_SPACING = 100d;
     public static final double CONTAINER_LEFTTORIGHT_BASE_SPACING = 15d;
     public static final double CONTAINER_LEFTTORIGHT_TOP_PADDING = 10d;
```
Comment (handreyrc) [inline packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/autolayout/elkjs/ELKUtils.java:56]: Hey @romartin,
Labels under the node are overlaping the container border, cosider increasing all padding properties to 25d.
There is a lost semicolon at the end of the block.

File: packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/autolayout/elkjs/ELKUtils.java [hunk 2]
```diff
@@ -51,7 +51,7 @@ public class ELKUtils {
...
     public static final double CONTAINER_LEFTTORIGHT_LEFT_PADDING = 10d;
```
Comment (handreyrc) [inline packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/autolayout/elkjs/ELKUtils.java]: Hey @romartin,

I've found some impovement points reviewing this PR:

Labels under the nodes are overlapping the container border, consider increasing all padding properties to 25d. There is a lost semicolon at the end of the block.

There is some inconsistency in the creation of Starting states.
- "start": "CheckInbox",  //works
- No start,  // even though it is not declared it's being inferred in other tools, we should consider doing the same.
- "start": { "stateName": "CheckInbox","schedule": {"cron": "0 0/15 * * * ?"}, //not supported

Docked nodes (Timer) are causing problems with the current auto-layout implementation. Consider changing this implementation if we are not supporting them.

Please, Could you take a look at the items above?
Thanks!

**Comment by tiagobento** [comment]
@romartin There's a formatting issue with this PR. Should be easily fixed with `yarn format` at the repository root.

**Comment by romartin** [comment]
Hey @domhanak 

Thanks for the first look at this!

I have properly pulled your javadoc and working on it, I'll add a commit in a while for those.

Also, about what you mention about the correlation of the (Java) domain classes with the spec, it's true that classes actually have not all fields as from the spec, but it's not an issue, let me explain: The beans are declared also as a JsType class, so once loading a workflow, all fields being loaded are part of the js object state anyway, even if no concrete field is present on the class, the js object itself (in the client side) contains all fields being parsed from the json.

Let you know that all those fields "missing", even if not being an issue, will be incremenally added, once we need to support them in some way, or if found any issues related, but in case a field from the spec has no "sense" for the editor, there is no real need for having it declared as a field in the class... Anyway, there actually exist a ticket [1] for reviewing all class fields and do some smoketests as well. This ticket is next step, so please keep in the loop! :)

[1] https://issues.redhat.com/browse/KOGITO-6975

Thanks!

File: packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/autolayout/elkjs/ELKUtils.java (hunk 2 above)
Comment (romartin) [reply packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/autolayout/elkjs/ELKUtils.java]: Hey @handreyrc 

Thanks for the updates! I properly changed margin values, also disabled timeout (docked) nodes for now. Adding a new commit with the changes...

Also about your issue with the `start` field, thanks for reporting into https://issues.redhat.com/browse/KOGITO-6975 , we'll tackle this ticket as next step!

**Comment by romartin** [comment]
Hey @domhanak @handreyrc 

Added commits for the changes requested. Ready for review again. 

Thanks!

**Comment by domhanak** [comment]
@romartin @handreyrc I just shared with you a document with full review result, please let me know your opinions on found issues!

View only link - https://docs.google.com/document/d/14ZDnGDXQ-i3HInGsmxCoHKaaEbWEpfXPYIVODOYv6mM/edit?usp=sharing

**Review by domhanak — CHANGES_REQUESTED** [review]
Some examples are not working properly, see my above comment and document with details.

**Comment by handreyrc** [comment]
Hey @romartin,

I tested your latest with the examples from the link bellow and we have some issues.

**EXAMPLES**: [Serverless Workflow Web Editor](https://serverlessworkflow.io/editor.html)

- Event-based decisions example : It breaks while unmarshalling. 
- Monitor jobs example : It breaks while unmarshalling. 
- Vet appointment example: The generated .sw file is broken and won't open.
- Monitor patient vitals example : The start state is not being created/inferred.
- Customer email example (async) : The start state is not being created/inferred.
- New item purchase example (compensation) : The start state is not being created/inferred.
- Check Inbox periodicall (cron) : The start state is not being created/inferred.

Could you please take a look?

Thanks!

**Comment by romartin** [comment]
Hey @domhanak @handreyrc 

Thanks a lot for your reviews guys! Jus let you know that I am working on analyzing all the reported issues and will provide fixes or feedback ASAP.

**Comment by romartin** [comment]
@handreyrc  I can properly load both examples "Event based decisions" and "Monitor jobs", no issues:

Screenshot from 2022-04-27 23-40-52
Screenshot from 2022-04-27 23-40-32

Which exception are you getting? Thx!

**Comment by romartin** [comment]
BTW @domhanak I really like the way you presented the PR review by using the document, so simple, so clear, and so easy to read and follow. Also with screenshots and links to examples. that's a great stuff!! Also we can easily collaborate in the doc, so maybe we should consider this way to do reviews in all (longer) PRs, all the team :) :+1:

**Comment by handreyrc** [comment]
Hey @romartin,

The  "Event based decisions" and "Monitor jobs" examples break when you try to download/unmarshall them.
This is the error I'm getting:
image

Thanks for checking these issues!

**Comment by romartin** [comment]
Hey @handreyrc 

Agree, those examples are not being properly serialized (when saving or downloading the workflow). This issue is because those contain swtich states. This is alrady being fixed by @Josephblt in the work for https://issues.redhat.com/browse/KOGITO-6975. Can you please confirm @Josephblt ?

This way @handreyrc , no other issues pending, you think it's ok for approving this PR?

Thanks guys!

**Comment by romartin** [comment]
Hey @handreyrc  - Fixed marshalling issues. Hope everything is ok now.

Hey @domhanak - Also fixed a few issues and others are out from scope. I have been adding comments in the review document for each item, I think we're ok to go now. Anyway let's properly discuss on next call all details!

Thanks!

**Review by tiagobento — APPROVED** [review]

**Comment by romartin** [comment]
NOTE : I have rebased from main

**Comment by romartin** [comment]
@handreyrc ready to test! Thanks!

**Comment by handreyrc** [comment]
Hey @romartin,

I've tested your latest and the Marshalling issues I reported above are fixed.  
We are good to move forward with this PR. It's approved.

 Thanks for your work here!

**Comment by romartin** [comment]
Hey @domhanak 

You were right!!! Some labels were still showing for some connectors, so just diabled! Thanks a lot for your feedback.

I have commited this, and it's a safe commit, I mean, you don't need to test again, changes are only about showing name or not for connectors (label).

**Review by domhanak — APPROVED** [review]
LGTM, most examples are working and the issues are reported for further enhancements.

I noticed some TODO's but as this is not used anywhere yet, we can ignore.

Thanks @romartin let's continue with the enhancements.

**Pull request merged**